java-checks/src/main/java/org/sonar/java/symexecengine/ExecutionState.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012 SonarSource
+ * dev@sonar.codehaus.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.symexecengine;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.sonar.java.locks.LockState;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import java.util.Map;
+import java.util.Set;
+
+public class ExecutionState {
+
+  private final IssuableSubscriptionVisitor check;
+  ExecutionState parent;
+  private Map<Symbol, Value> valuesBySymbol = Maps.newHashMap();
+
+  public ExecutionState(ExecutionState executionState) {
+    this.parent = executionState;
+    this.check = executionState.check;
+  }
+
+  public ExecutionState(IssuableSubscriptionVisitor check) {
+    this.check = check;
+  }
+
+  public ExecutionState merge(ExecutionState executionState) {
+    for (Map.Entry<Symbol, Value> entry : executionState.valuesBySymbol.entrySet()) {
+      Symbol symbol = entry.getKey();
+      Value currentValue = getValue(symbol);
+      Value valueToMerge = entry.getValue();
+      if (currentValue != null) {
+        currentValue.state = currentValue.state.merge(valueToMerge.state);
+        valuesBySymbol.put(symbol, currentValue);
+      } else {
+        // possible way to solve the problem of variable defined in outer Execution state. lockedOccurencesBySymbol.put(symbol,
+        // occurenceToMerge);
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)This block of commented-out lines of code should be removed.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)